src/gtest/test_paymentdisclosure.cpp

```diff
@@ -0,0 +1,208 @@
+        std::vector<unsigned char> vch(&buffer[0], &buffer[0] + 32);
+        uint256 joinSplitPrivKey = uint256(vch);
+
+        // Create payment disclosure key and info data to store in test database
+        size_t js = random_uint256().GetCheapHash() % std::numeric_limits<size_t>::max();
+        uint8_t n = random_uint256().GetCheapHash() % std::numeric_limits<uint8_t>::max();
+        PaymentDisclosureKey key { random_uint256(), js, n};
+        PaymentDisclosureInfo info;
+        info.esk = random_uint256();
+        info.joinSplitPrivKey = joinSplitPrivKey;
+        info.zaddr = libzcash::SpendingKey::random().address();
+        ASSERT_TRUE(mydb.Put(key, info));
+
+        // Retrieve info from test database into new local variable and test it matches
+        PaymentDisclosureInfo info2;
+        ASSERT_TRUE(mydb.Get(key, info2));
+        ASSERT_EQ(info, info2);
+
+        // Modify this local variable and confirm it no longer matches
+        info2.esk = random_uint256();
+        info2.joinSplitPrivKey = random_uint256();
+        info2.zaddr = libzcash::SpendingKey::random().address();        
+        ASSERT_NE(info, info2);
+
+        // Using the payment info object, let's create a dummy payload
+        PaymentDisclosurePayload payload;
+        payload.version = PAYMENT_DISCLOSURE_VERSION_EXPERIMENTAL;
+        payload.esk = info.esk;
+        payload.txid = key.hash;
+        payload.js = key.js;
+        payload.n = key.n;
+        payload.message = "random-" + boost::uuids::to_string(uuidgen());   // random message
+        payload.zaddr = info.zaddr;
+
+        // Serialize and hash the payload to generate a signature
+        uint256 dataToBeSigned = SerializeHash(payload, SER_GETHASH, 0);
+
+        // Compute the payload signature
+        unsigned char payloadSig[64];
+        if (!(crypto_sign_detached(&payloadSig[0], NULL,
+            dataToBeSigned.begin(), 32,
+            &buffer[0] // buffer containing both private and public key required
+            ) == 0))
+        {
+            throw std::runtime_error("crypto_sign_detached failed");
+        }
+
+        // Sanity check
+        if (!(crypto_sign_verify_detached(&payloadSig[0],
+            dataToBeSigned.begin(), 32,
+            joinSplitPubKey.begin()
+            ) == 0))
+        {
+            throw std::runtime_error("crypto_sign_verify_detached failed");
+        }
+
+        // Convert signature buffer to boost array
+        boost::array<unsigned char, 64> arrayPayloadSig;
+        memcpy(arrayPayloadSig.data(), &payloadSig[0], 64);
+
+        // Payment disclosure blob to pass around
+        PaymentDisclosure pd = {payload, arrayPayloadSig};
+
+        // Test payment disclosure constructors
+        PaymentDisclosure pd2(payload, arrayPayloadSig);
+        assert(pd == pd2);
+        PaymentDisclosure pd3(joinSplitPubKey, key, info, payload.message);
+        assert(pd == pd3);
```

Comment:
``ASSERT_EQ(pd, pd3);``

Reply:
Similarly for the other asserts below in this test.

Reply:
ok